05-dijkstras/06-dijkstras-vs-negative-graphs.md

```diff
@@ -0,0 +1,122 @@
+
+```py
+current: D 
+
+visited nodes: 
+    - A
+    - D 
+
+Priority queue:
+    - B
+
+costs/distances: 
+- A: 0 
+- B: 2
+- C: ∞ 
+- D: 1
+
+```
+
+![Dijkstra vs.Negative Graph-5](images/negative-5.png)
+
+Step 4. We now look at the next node at the top of our priority queue, Node B (cost of 2). We will estimate the accumulated cost of reaching Node B’s neighboring nodes, based on the cost we calculated for reaching Node B, plus the cost to reach its neighbors. Node B has a single unvisited neighbor, Node C, and the edge to reach it has a cost of -100. The -100 cost of that edge to C, added to the cost it took to reach Node B (2) results in an estimated cost to Node C of -98. We can update our cost for Node C with our new found estimate, and since we are now done visiting Node B, we can add Node B to our visited nodes. Finally, we add Node C to our priority queue.
+
+```py
+current: B 
+
+visited nodes: 
+    - A
+    - D
+    - B 
+
+Priority queue:
+    - C
+
+costs/distances: 
+- A: 0 
+- B: 2
+- C: -98 
+- D: 1
+```
+
+![Dijkstra vs.Negative Graph-6](images/negative-6.png)
+
+Step 5. The next node in our priority queue is Node C. Like before, we will take a look at all of Node C’s unvisited neighbors and calculate the actual cost of visiting. As we can notice, Node C actually does not have any unvisited nodes and therefore leaves us with nothing to update.
+This brings our algorithm to an end and leaves us with the following result. 
+
+```py
+visited nodes: 
+    - A
+    - D
+    - B
+    - C
+
+costs/distances: 
+- A: 0 
+- B: 2
+- C: -98 
+- D: 1
+```
+But wait, we can clearly see that the most least costly path from Node A to Node D is through Nodes B and C! That is correct, however since Dijkstra’s Algorithm operates under the assumption that each local choice we make is the most optimal choice, Node D was visited early in the traversal when it appeared to be the next lowest cost node to visit after Node A. The cost we estimated at the time it was visited became its calculated cost. Even though we found a lower cost path later, the negative edge causes the greedy assumption "the path we find first will be the cheapest" to become violated. Because Node D had already been visited, the edge to Node D along this later, cheaper path would not be considered.This is why Dijkstra’s Algorithm can possibly fail when implemented on graphs with negative weights. To circumvent this limitation, we encourage you to follow your own curiosity and research the A* Search and Bellman Ford Algorithms.
```

Comment:
Just "least", not "most least".
Missing space
Consider splitting the FYC note into it's own paragraph

```suggestion
But wait, we can clearly see that the least costly path from Node A to Node D is through Nodes B and C! That is correct, however since Dijkstra’s Algorithm operates under the assumption that each local choice we make is the most optimal choice, Node D was visited early in the traversal when it appeared to be the next lowest cost node to visit after Node A. The cost we estimated at the time it was visited became its calculated cost. Even though we found a lower cost path later, the negative edge causes the greedy assumption "the path we find first will be the cheapest" to become violated. Because Node D had already been visited, the edge to Node D along this later, cheaper path would not be considered. This is why Dijkstra’s Algorithm can possibly fail when implemented on graphs with negative weights.

To circumvent the negative weight limitation, we encourage you to follow your own curiosity and research the A* Search and Bellman Ford Algorithms.
```